### ISSUE TITLE
Don't run 100 tests if there aren't 100 possible inputs

### ISSUE THREAD

**Issue by ocharles** [hedgehogqa/haskell-hedgehog#137]
We have the following:

```haskell
databaseTests :: TestDatabase -> PropertyT IO ()
databaseTests testDatabase = do
  -- Setup
  factory <- forAll Gen.enumBounded

  -- Expectation
  let expected = []

  test $ do
    liftIO (putStrLn "Ding")
    actual <- liftIO (runNoLoggingT (runPool (getKitBoxesForFactory factory) (testConnPool testDatabase)))
    actual === expected
```

We only `forAll` a single `Bounded` generator, which only has two possible values. However, when running this, the test is executed 100 times - which is obviously pointless.

It'd be nice to know this ran exhaustively and didn't repeat the same test more than once - which is sufficient criteria to know that only the two possible branches were executed.

**Comment by tmcgilchrist** [comment]
Have a look at the withTests function that should do what you want. “Set
the number times a property should be executed before it is considered
successful.”

**Comment by ocharles** [comment]
Yes, but I shouldn't need to specify it - it's statically available to find
out 2 tests could be sufficient. The default of 100 is good to explore a
big search space, but if <100 could exhaust the search, then it should.
Right now it's possible 100 tests wouldn't even hit all two cases, as it's
sampled multiple times.

**Comment by thumphries** [comment]
> Have a look at the withTests function

It would be nice to be able to figure this out automatically. I'm not sure it'd ever be possible to ensure every value had been generated given the current structure; it might involve introducing a new type of exhaustive generator or generation strategy. Not sure if that would make sense when combined with other generators in the same property.

I wonder if some shorthand for testing over bounded / enumerable domains would be worthwhile. Something like `Enum a => Bounded a => (a -> Property) -> Property` that exhausts the domain and stops.

**Comment by ocharles** [comment]
Well if we consider tests as walking through paths in a tree, we could add
the invariant that we never follow the same path twice. Hence in this
scenario after 2 runs we'd have no more possible paths and we'd be done.

**Comment by albertnetymk** [comment]
> after 2 runs we'd have no more possible paths and we'd be done.

That would require tight coupling between the test runner and the generator. From the generator's perspective, it only knows how to "generate" a value. Even when the candidate pool has size 1, I can still ask it for a value *twice*, e.g. `Gen.list (Range.singleton 2) (Gen.constant 1)`. Should the generator refuse my second query just because it can't provide anything new? I don't think so.

The OP's code snippet uses the generator directly, but in a more general setting, generators could be used to construct other generators.
